sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md

```diff
@@ -0,0 +1,113 @@
+# SYCL(TM) Proposal: New device descriptors for Intel GPUs
+
+**IMPORTANT**: This specification is a draft.
+
+**NOTE**: Khronos(R) is a registered trademark and SYCL(TM) is a trademark of the Khronos Group, Inc.
+
+New device descriptors will be added to provide access to low-level hardware details about Intel GPU devices.  This information will be useful to developers tuning specifically for those devices.
+
+This proposal details what is required to provide this information as a SYCL extensions.
+
+## Feature Test Macro ##
+
+The same Feature Test Macro will be used for the new device descriptor support.  It will be defined as:
+
+    #define SYCL_EXT_INTEL_GPU_DEVICE_INFO 1
+
+
+
+# Intel GPU PCI Address #
+
+A new device descriptor will be added which will provide the PCI address of an Intel GPU in BDF format.  BDF format contains the address as: `domain:bus:device.function`.
+
+This support can only be provided when using the Level Zero backend \(BE\).  This is the default for Intel GPUs.  The OpenCL BE does not provide the PCI address at this time.
+
+
+## Device Information Descriptors ##
+
+| Device Descriptors | Return Type | Description |
+| ------------------ | ----------- | ----------- |
+| info\:\:device\:\:ext\_intel\_gpu\_pci\_address | std\:\:string | For Level Zero BE, returns the PCI address in BDF format: `domain:bus:device.function`.|
+
+
+## Aspects ##
+
+A new aspect, ext\_intel\_gpu\_pci\_address, will be added.
+
+## Error Condition ##
+
+The function device\:\:get_info\(\) will return an empty string if the device does not support aspect\:\:ext\_intel\_gpu\_pci\_address.
+
+
+## Example Usage ##
+
+The PCI address can be obtained using the standard get\_info() interface.
+
+    if (dev.has(aspect::ext_intel_gpu_pci_address) {
+      auto BDF = dev.get_info<info::device::ext_intel_gpu_pci_address>();
+    }
+
+
+
+# Intel GPU Execution Unit SIMD Width #
+
+A new device descriptor will be added which will provide the physical SIMD width of an execution unit on an Intel GPU.  This data will be used to calculate the computational capabilities of the device.
+
+This support will only be provided when using the Level Zero backend \(BE\).  This is the default for Intel GPUs.  The OpenCL BE does not provide the physical SIMD width at this time.
+
+## Device Information Descriptors ##
+
+| Device Descriptors | Return Type | Description |
+| ------------------ | ----------- | ----------- |
+| info\:\:device\:\:ext\_intel\_gpu\_eu\_simd\_width | uint32\_t| Returns the physical SIMD width of the  execution unit (EU).|
+
+
+## Aspects ##
+
+A new aspect, ext\_intel\_gpu\_eu\_simd\_width, will be added.
+
+
+## Error Condition ##
+
+The function device\:\:get_info\(\) will return PI\_INVALID\_VALUE if the device does not support aspect\:\:ext\_intel\_gpu\_eu\_simd\_width.
```

Comment:
This seems like a bad idea for two reasons.  First, the value `PI_INVALID_VALUE` isn't defined in the SYCL specification nor is it defined in this extension.

However, the more important concern is that callers won't be able to distinguish this return value from a real SIMD width.  They're both just numbers, and the numerical value of `PI_INVALID_VALUE` might just happen to be the same as a valid SIMD width.

I think it would be better if all these queries threw an `errc::invalid_object` exception if they are called for an unsupported device.  Note that there is already a precedent for throwing an exception when calling get_info() with illegal input.  The existing descriptor `info::device::parent_device` currently throws `errc::invalid_object` if it is called for a device that is not a sub-device.

Reply:
It will now throw an exception when the support is not present.

Reply:
Thanks for adding that, but the spec should say which exception is thrown.  I'd suggest wording something like:

```
If the device does not support aspect\:\:ext\_intel\_gpu\_pci\_address, `get_info()` throws
an `exception` with the `errc::invalid_object` error code.
```

(I'm not sure of the markdown formatting, so you should check that.)

Reply:
I regret being so low-level in my original posting.  I have updated this statement to be more general.  Hopefully this will satisfy your concerns.